src/json-patch.ts

```diff
@@ -256,92 +266,280 @@ module jsonpatch {
+  * Escapes a json pointer path
+  * @param path The raw pointer
+  * @return the Escaped path
+  */
+  export function escapePathComponent(path: string): string {
+    if (path.indexOf('/') === -1 && path.indexOf('~') === -1) return path;
+    return path.replace(/~/g, '~0').replace(/\//g, '~1');
+  }
+  /**
+   * Unescapes a json pointer path
+   * @param path The escaped pointer
+   * @return The unescaped path
+   */
+  export function unescapePathComponent(path: string): string {
+    return path.replace(/~1/g, '/').replace(/~0/g, '~');
+  }
+
+  /**
+   * Retrieves a value from a JSON document by a JSON pointer.
+   * Returns the value.
+   *
+   * @param document The document to get the value from
+   * @param pointer an escaped JSON pointer
+   * @return The retrieved value
    */
-  export function apply(tree: any, patches: Patch<any>[], validate?: boolean): any[] {
-    var results = new Array(patches.length)
-      , p = 0
-      , plen = patches.length
-      , patch
-      , key;
-    while (p < plen) {
-      patch = patches[p];
-      p++;
-      // Find the object
-      var path = patch.path || "";
-      var keys = path.split('/');
-      var obj = tree;
-      var t = 1; //skip empty element - http://jsperf.com/to-shift-or-not-to-shift
-      var len = keys.length;
-      var existingPathFragment = undefined;
-
-      while (true) {
+  export function getValueByPointer(document: any, pointer: string): any {
+    let pathSegments = [];
+    if (pointer) { //empty string evaluates to false
+      pathSegments = pointer.substring(1).split(/\//).map(jsonpatch.unescapePathComponent);
+    }
+    for (let i = 0; i < pathSegments.length; i++) {
+      let subPropertyKey = pathSegments[i];
+      if (!(subPropertyKey in document)) {
+        return undefined;
+      }
+      document = document[subPropertyKey];
+    }
+    return document;
+  }
+  /**
+   * Apply a single JSON Patch Operation on a JSON document.
+   * Returns {newDocument, result} of the operation.
+   *
+   * @param document The document to patch
+   * @param operation The operation to apply
+   * @return `{newDocument, result}` after the operation
+   */
+  export function applyOperation<T>(document: T, operation: Operation): OperationResult<T>;
+
+  /**
+   * Apply a single JSON Patch Operation on a JSON document.
+   * Returns the {newDocument, result} of the operation.
+   *
+   * @param document The document to patch
+   * @param operation The operation to apply
+   * @param validateOperation `false` is without validation, `true` to use default jsonpatch's validation, or you can pass a `validateOperation` callback to be used for validation.
+   * @param mutateDocument Whether to mutate the original document or clone it before applying
+   * @return `{newDocument, result}` after the operation
+   */
+  export function applyOperation<T>(document: any, operation: Operation, validateOperation?: boolean | Function, mutateDocument?: boolean): OperationResult<T>;
```

Comment:
Just tried it out locally - the signature above that also needs to go, since we now have `applyReducer` for the reduce use case we don't need it anymore.